trust-quorum/src/task.rs

```diff
@@ -93,6 +198,114 @@ impl NodeTaskHandle {
         &self.baseboard_id
     }
 
+    /// Initiate a trust quorum reconfiguration at this node
+    pub async fn reconfigure(
+        &self,
+        msg: ReconfigureMsg,
+    ) -> Result<(), NodeApiError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(NodeApiRequest::Reconfigure { msg, responder: tx })
+            .await?;
+        rx.await??;
+        Ok(())
+    }
+
+    /// Initiate an LRTQ upgrade at this node
+    pub async fn upgrade_from_lrtq(
+        &self,
+        msg: LrtqUpgradeMsg,
+    ) -> Result<(), NodeApiError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(NodeApiRequest::LrtqUpgrade { msg, responder: tx })
+            .await?;
+        rx.await??;
+        Ok(())
+    }
+
+    /// Return the status of this node if it is coordinating the `Prepare` phase
+    /// of a reconfiguration or LRTQ upgrade. Return `Ok(None)` or an error
+    /// otherwise.
+    pub async fn coordinator_status(
+        &self,
+    ) -> Result<Option<CoordinatorStatus>, NodeApiError> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(NodeApiRequest::CoordinatorStatus { responder: tx })
+            .await?;
+        let res = rx.await?;
+        Ok(res)
+    }
+
+    /// Load the rack secret for the given epoch
+    ///
+    /// This can block for an indefinite period of time before returning
+    /// and depends on availability of the trust quorum.
+    pub async fn load_rack_secret(
+        &self,
+        epoch: Epoch,
+    ) -> Result<ReconstructedRackSecret, NodeApiError> {
+        loop {
+            let (tx, rx) = oneshot::channel();
+            self.tx
+                .send(NodeApiRequest::LoadRackSecret { epoch, responder: tx })
+                .await?;
+            if let Some(rack_secret) = rx.await?? {
+                return Ok(rack_secret);
+            };
+
+            // The task returns immediately with `None` if the secret is still
+            // being loaded. We must therefore retry.
+            sleep(LOAD_RACK_SECRET_RETRY_TIMEOUT).await;
+        }
+    }
+
+    /// Return `Ok(true)` if the configuration has committed, `Ok(false)` if
+    /// it hasn't committed yet, or an error otherwise.
+    ///
+    /// Nexus will retry this operation and so we should only try once here.
+    /// This is in contrast to operations like `load_rack_secret` that are
+    /// called directly from sled agent.
+    pub async fn prepare_and_commit(
```

Comment:
Could you return a two-valued enum here rather than `bool`?

What is the "otherwise" in "return an error otherwise" here? Just send and receive errors or something else?

Also since this doesn't loop I'd consider calling this `try_prepare_and_commit`.

Reply:
Totally. As I put that bool there, I knew you'd ask for this. Next time I'll do it **before** you ask ;)